### PR TITLE
[PEP 741] gh-107954: Add PyConfig_Get() function

### PR DESCRIPTION
**Pull request by vstinner**
Add PyConfig_Get() and PyConfig_GetInt() functions to get the current Python configuration.

_PyConfig_AsDict() now converts PyConfig.xoptions as a dictionary.

**Discussion:** https://discuss.python.org/t/fr-allow-private-runtime-config-to-enable-extending-without-breaking-the-pyconfig-abi/18004

* Issue: gh-107954

----
:books: Documentation preview :books:: https://cpython-previews--112609.org.readthedocs.build/

### CODE REVIEW

**Comment by vstinner** [comment]
This change doesn't add the new functions to the limited C API yet. They can be added in a follow-up PR.

File: Doc/c-api/init_config.rst [hunk 1]
```diff
@@ -1602,6 +1602,53 @@ customized Python always running in isolated mode using
...
+====================================
+
+Get a configuration option where *name* is the name of a :c:type:`PyConfig`
+member.
+
+Some options are read from the :mod:`sys` attributes. For example, the option
+``"argv"`` is read from :data:`sys.argv`.
+
+
+.. c:function:: int PyConfig_Get(const char *name, PyObject **value)
```
Comment (zooba) [inline Doc/c-api/init_config.rst]: I assume this has to be run with the GIL held and after initialization?

File: Doc/c-api/init_config.rst [hunk 2]
```diff
@@ -1602,6 +1602,53 @@ customized Python always running in isolated mode using
...
+
+   The object type depends on the configuration option. It can be:
+
+   * ``int``
+   * ``str``
+   * ``list[str]``
+   * ``dict[str, str]``
+
+   * Return ``0`` and set *\*value* on success.
+   * Raise an exception and return ``-1`` on error.
```
Comment (zooba) [inline Doc/c-api/init_config.rst]: If the result is an error or a `PyObject*`, then the return type can just be `PyObject *` and NULL on error. (And I guess worth mentioning that https://github.com/capi-workgroup/api-evolution/issues/13 is still a discussion and not a guideline yet.)

File: Doc/c-api/init_config.rst (hunk 1 above)
Comment (vstinner) [reply Doc/c-api/init_config.rst]: Right, I completed the doc.

File: Doc/c-api/init_config.rst (hunk 2 above)
Comment (vstinner) [reply Doc/c-api/init_config.rst]: Oops sorry, I changed the API multiple times, there were many back and forth. At the end, I forgot to update the doc! Fixed. The function does return a Python object (`PyObject*`), or NULL (and set an exception) on error.

**Review by zooba — APPROVED** [review]
LGTM

**Comment by vstinner** [comment]
I created an issue for the C API Working Group to add this C API: https://github.com/capi-workgroup/decisions/issues/3

**Comment by vstinner** [comment]
I expected such API (two functions) to be simple to agree on, but apparently, it's more completed than expected. @encukou now [suggests writing a PEP](https://github.com/capi-workgroup/decisions/issues/3#issuecomment-1862551858) for adding these two functions. I'm already struggling to get a consensus on my [PEP 737 – Unify type name formatting](https://discuss.python.org/t/pep-737-unify-type-name-formatting/39872). I don't have the bandwidth to fight for too many APIs in parallel. I prefer to close the PR for now. I may reopen it when I will have more bandwidth to look into this topic again.

**Comment by vstinner** [comment]
Since this PR is closed, I created a new PR to implement PEP 741: https://github.com/python/cpython/pull/123472

**Pull request closed without merging**